src/transformStateWithClones.js

```diff
@@ -7,7 +7,35 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const ACTIONS = [...actions];
+  let newState = { ...state };
+  const result = [];
+
+  for (const action of ACTIONS) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(newState, action.extraData);
+        result.push({ ...newState });
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete newState[key];
+        }
+        result.push({ ...newState });
+        break;
+
+      case 'clear':
+        newState = {};
+        result.push({ ...newState });
+        break;
+
+      default:
+        throw new Error('Unknown action type');
```

Comment:
The current implementation throws an error for unknown action types. While this is a valid approach, consider whether this behavior is desired in your application context. If unknown actions should be ignored or logged instead of throwing an error, you might want to adjust this logic.